app/services/topology_service.rb

```diff
@@ -1,22 +1,8 @@
 class TopologyService
-  class << self
-    attr_reader :provider_class
-  end
-
-  def initialize(provider_id)
-    @providers = retrieve_providers(provider_id)
-  end
-
-  def retrieve_providers(provider_id = nil)
-    if provider_id
-      retrieve_entity(provider_id)
-    else  # provider id is empty when the topology is generated for all the providers together
-      self.class.provider_class.all
-    end
-  end
-
-  def retrieve_entity(entity_id)
-    self.class.provider_class.where(:id => entity_id)
+  def initialize(provider_id = nil)
+    provider_class = self.class.instance_variable_get(:@provider_class)
+    # If the provider ID is not set, the topology needs to be generated for all the providers
+    @providers = provider_id ? provider_class.where(:id => provider_id) : provider_class.all
```

Comment:
I'd like to use a new method introduces in #1137:

```ruby
@providers = if provider_id
               find_records_with_rbac(provider_class, provider_id)
             else
               Rbac.filtered(provider_class.all)
             end
```

I'll create a PR once it's merged.